flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTaskTest.java

```diff
@@ -856,4 +887,41 @@ private void output(String record) {
             output.collect(new StreamRecord<>(record));
         }
     }
+
+    /**
+     * This source sleeps a little bit before processing cancellation and records whether it was
+     * interrupted by the {@link SourceStreamTask} or not.
+     */
+    private static class WasInterruptedTestingSource implements SourceFunction<String> {
+        private static final int MAX_POST_CANCEL_ITERATIONS = 100;
+        private static final long serialVersionUID = 1L;
+        private volatile boolean running;
+
+        public static boolean wasInterrupted;
+
+        @Override
+        public void run(SourceContext<String> ctx) throws Exception {
+            wasInterrupted = false;
+            try {
+                int postCancelIterations = 0;
+                while (running || postCancelIterations < MAX_POST_CANCEL_ITERATIONS) {
+                    Thread.sleep(1);
+                    if (!running) {
+                        postCancelIterations++;
+                    }
+                }
+            } catch (InterruptedException e) {
+                wasInterrupted = true;
```

Comment:
I'm afraid there is very small probability that the thread will be interrupted after the last sleep.
So I'd check the interrupted status explicitly (in addition to this `catch`).

Reply:
The tested window if interruption has happened or not is 100ms (10^-1 seconds). After the sleep we have only a couple of instructions to perform (exiting the loop). So adding `interrupted` such check would extend this window by  something 1 micro second (10^-6 s) (assuming we would need to check volatiles) and 1 nano second (10^-9 s). 

That's a difference of 5 to 8 orders of magnitudes, I don't think extra complexity/more code is worth for checking that.

Reply:
I agree that the probability is very low (also such check won't be needed if we don't loop in the source and ony `await` on some latch).